Aos 2 approximate unbounded efif

efifogel opened the pull request:
## Summary of Changes

Enhanced the drawing capabilities to draw arrangements induced by unbounded curves. This small feature consists of (i) additional draw() overloads for arrangements that also accepts a bounding box, and (ii) an additional requirement from the geometry traits for an operator that approximates a curve and also accepts a bounding box. 

## Release Management

* Affected package(s): Arrangement_on_surface_2
* Issue(s) solved (if any): 
* Feature/Small Feature (if any): [Aos2 Unbounded Drawing](https://cgalwiki.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Aos2_Unbounded_Drawing)
* Link to compiled documentation (obligatory for small feature) [doc_output](https://www.cs.tau.ac.il/~efif/doc_output12/Arrangement_on_surface_2/)
* License and copyright ownership: TAU

efifogel committed:
Added draw() overloads that accept a bounding box and fixed the Approximate_2 concept documentation

efifogel committed:
Introduced ApproximateUnbounded_2 concept

efifogel committed:
Added functionality to draw arrangements induced by unbounded curves

efifogel committed:
Enhanced the description of the draw() overloads

efifogel committed:
Fixed typo in Bbox type

efifogel committed:
Add Approximate_2

efifogel committed:
Guard for unbounded faces

efifogel committed:
Added missing include CGAL/Cartesian.h

efifogel committed:
Enhanced

efifogel committed:
Updated concepts

afabri commented on Arrangement_on_surface_2/doc/Arrangement_on_surface_2/Concepts/AosTraits--ApproximateUnbounded_2.h:
```diff
@@ -0,0 +1,49 @@
+   *        approximates the curve.
+   * \param oi An output iterator for the output containers.
+   * \param bbox the bounding box.
+   * \param l2r A Boolean flag that indicates whether the curve direction is
```
```suggestion
   * \param l2r A %Boolean flag that indicates whether the curve direction is
```

efifogel replied:
Why is it needed? "Boolean" does not create an undesired link.
I've searched for "Boolean" in txt files and found 233 occurrences out of which only 2 have the '%' as a prefix and only in the Polygon_repair/doc/Polygon_repair/Polygon_repair.txt file.
I would guess that the `%` character should be removed. Am I missing something?

afabri replied:
I had the wrong idea that we have a type `CGAL::Boolean`. Sorry for the noise.

afabri replied:
And yes, it can be removed.

afabri commented on Arrangement_on_surface_2/doc/Arrangement_on_surface_2/Concepts/AosTraits--ApproximatePoint_2.h:
```diff
@@ -0,0 +1,28 @@
+   * `p`'s \f$y\f$-coordinate (if `i == 1`).
+   * \pre `i` is either 0 or 1.
+   */
+  CGAL::Approximate_number_type operator()(AosTraits::Point_2 p, int i);
```
Do you really want to introduce `CGAL::Approximate_number_type`???

afabri commented on Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h:
```diff
@@ -1517,9 +1518,24 @@ class Arr_linear_traits_2 : public Kernel_ {
   //@{
-  typedef double                          Approximate_number_type;
+  typedef double                                        Approximate_number_type;
+  typedef CGAL::Cartesian<Approximate_number_type>      Approximate_kernel;
```
Is `Cartesian` better suited than `Simple_cartesian`?

afabri commented:
For the approximation business, wouldn't it be reasonable to use the [`CGAL::Cartesian_coverter`](https://doc.cgal.org/latest/Kernel_23/classCGAL_1_1Cartesian__converter.html)?

efifogel committed:
Removed erroneous CGAL:: (from CGAL::Approximate_...)

efifogel commented:
Yes, it makes sense to use it in the implementation.
I think that the following traits classes can benefit from this:

   - Arr_circle_segment_traits_2.h (maybe)
   - Arr_conic_traits_2.h (maybe)
   - Arr_geodesic_arc_on_sphere_traits_2.h (maybe)
   - Arr_linear_traits_2.h
   - Arr_non_caching_segment_basic_traits_2.h
   - Arr_segment_traits_2.h

I'll go ahead and adapt it.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Fri, 22 Aug 2025 at 13:20, Andreas Fabri ***@***.***>
wrote:

> *afabri* left a comment (CGAL/cgal#9036)
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3213865581>
>
> For the approximation business, wouldn't it be reasonable to use the
> CGAL::Cartesian_coverter
> <https://doc.cgal.org/latest/Kernel_23/classCGAL_1_1Cartesian__converter.html>
> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3213865581>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOBUVIIGYFJ5FQ32W4D3O3VH5AVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMJTHA3DKNJYGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

efifogel commented:
I believe that the PR is now ready for integration.

The two additional changes:
1. Using `Simple_cartesian` instead of `Cartesian`, and
2.  Using [GAL::Cartesian_coverter](https://doc.cgal.org/latest/Kernel_23/classCGAL_1_1Cartesian__converter.html)

will be done in a follow up PR.
I'd like to keep this one minimal, because there is another PR pending (implemented as part of GSoC), which relies on this one and already contains the changes of this one, and I would like to avoid conflicts as much as possible.

efifogel committed:
Fixed drawing with a bounding box

efifogel commented:
This was fixed during the day today
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Wed, 17 Sept 2025 at 16:35, Sebastien Loriot ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h
> <https://github.com/CGAL/cgal/pull/9036#discussion_r2355548056>:
>
> > +      const auto& trg = (l2r) ? max_vertex(xcv) : min_vertex(xcv);
> +      auto xs = CGAL::to_double(src.x());
> +      auto ys = CGAL::to_double(src.y());
> +      auto xt = CGAL::to_double(trg.x());
> +      auto yt = CGAL::to_double(trg.y());
> +      *oi++ = Approximate_point_2(xs, ys);
> +      *oi++ = Approximate_point_2(xt, yt);
> +      return oi;
> +    }
> +
> +    /*! obtains an approximation of an \f$x\f$-monotone curve.
> +     */
> +    template <typename OutputIterator>
> +    OutputIterator operator()(const X_monotone_curve_2& xcv, double /* error */,
> +                              OutputIterator oi, const Bbox_2& bbox,
> +                              bool l2r = true) const
>
> warning: l2r is not used!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#pullrequestreview-3234501555>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOCLNFUEDACHPLRXZG33TFPRHAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMZUGUYDCNJVGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sloriot commented on Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h:
```diff
@@ -1533,10 +1549,102 @@ class Arr_linear_traits_2 : public Kernel_ {
+      if (! result) return oi;
+
+      if (const auto* res_seg = std::get_if<Approx_seg>(&*result)) {
+        *oi++ = l2r ? res_seg->min() : res_seg->max();
+        *oi++ = l2r ? res_seg->max() : res_seg->min();
```
```suggestion
        *oi++ = l2r ? (res_seg->min)() : (res_seg->max)();
        *oi++ = l2r ? (res_seg->max)() : (res_seg->min)();
```

efifogel commented:
Is this really necessary nowadays? I thought that we passed that.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Thu, 18 Sept 2025 at 08:17, Sebastien Loriot ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h
> <https://github.com/CGAL/cgal/pull/9036#discussion_r2357542096>:
>
> > +        *oi++ = l2r ? res_seg->min() : res_seg->max();
> +        *oi++ = l2r ? res_seg->max() : res_seg->min();
>
> ⬇️ Suggested change
>
> -        *oi++ = l2r ? res_seg->min() : res_seg->max();
> -        *oi++ = l2r ? res_seg->max() : res_seg->min();
> +        *oi++ = l2r ? (res_seg->min)() : (res_seg->max)();
> +        *oi++ = l2r ? (res_seg->max)() : (res_seg->min)();
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#pullrequestreview-3237318241>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOC4RRBLQK5XZ5AYEYL3TI57NAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMZXGMYTQMRUGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

afabri commented:
> Is this really necessary nowadays? I thought that we passed that. ____ _ ____ _ /_____/_) o /__________ __ // (____ ( ( ( (_/ (_/-(-'_(/ _/
> […](#)
> On Thu, 18 Sept 2025 at 08:17, Sebastien Loriot ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h <[#9036 (comment)](https://github.com/CGAL/cgal/pull/9036#discussion_r2357542096)>: > + *oi++ = l2r ? res_seg->min() : res_seg->max(); + *oi++ = l2r ? res_seg->max() : res_seg->min(); ⬇️ Suggested change - *oi++ = l2r ? res_seg->min() : res_seg->max(); - *oi++ = l2r ? res_seg->max() : res_seg->min(); + *oi++ = l2r ? (res_seg->min)() : (res_seg->max)(); + *oi++ = l2r ? (res_seg->max)() : (res_seg->min)(); — Reply to this email directly, view it on GitHub <[#9036 (review)](https://github.com/CGAL/cgal/pull/9036#pullrequestreview-3237318241)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABVBNOC4RRBLQK5XZ5AYEYL3TI57NAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMZXGMYTQMRUGE> . You are receiving this because you authored the thread.Message ID: ***@***.***>

It is needed as soon as user code does `#include <windows.h>`.  I sometimes see also `#define NOMINMAX`  as a remedy, but that may not be desired from the user point of view.

efifogel commented:
Which version of MSVC (causes a problem when including <windows.h>)?

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Thu, 18 Sept 2025 at 10:30, Andreas Fabri ***@***.***>
wrote:

> *afabri* left a comment (CGAL/cgal#9036)
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3305822765>
>
> Is this really necessary nowadays? I thought that we passed that. ____ _
> ____ _ /_____/*) o /*_________ __ // (____ ( ( ( (*/ (*/-(-'_(/ _/
> … <#m_287528763086160393_>
> On Thu, 18 Sept 2025 at 08:17, Sebastien Loriot *@*.*> wrote: @.**
> commented on this pull request. ------------------------------ In
> Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h <#9036
> (comment) <https://github.com/CGAL/cgal/pull/9036#discussion_r2357542096>>:
> > + *oi++ = l2r ? res_seg->min() : res_seg->max(); + *oi++ = l2r ?
> res_seg->max() : res_seg->min(); ⬇️ Suggested change - *oi++ = l2r ?
> res_seg->min() : res_seg->max(); - *oi++ = l2r ? res_seg->max() :
> res_seg->min(); + oi++ = l2r ? (res_seg->min)() : (res_seg->max)(); + oi++
> = l2r ? (res_seg->max)() : (res_seg->min)(); — Reply to this email
> directly, view it on GitHub <#9036 (review)
> <https://github.com/CGAL/cgal/pull/9036#pullrequestreview-3237318241>>, or
> unsubscribe
> https://github.com/notifications/unsubscribe-auth/ABVBNOC4RRBLQK5XZ5AYEYL3TI57NAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMZXGMYTQMRUGE
> <https://github.com/notifications/unsubscribe-auth/ABVBNOC4RRBLQK5XZ5AYEYL3TI57NAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTEMZXGMYTQMRUGE>
> . You are receiving this because you authored the thread.Message ID: @.*>
>
> It is needed as soon as user code does #include <windows.h>. I sometimes
> see also #define NOMINMAX as a remedy, but that may not be desired from
> the user point of view.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3305822765>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOBQMEQ6USI7ZT6TEEL3TJNRZAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMBVHAZDENZWGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

efifogel committed:
Pacify MSVC (min/max issue)

efifogel commented:
ok, I fixed it.
I guess they will never fix it because of backward compatibility.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Thu, 18 Sept 2025 at 12:47, Andreas Fabri ***@***.***>
wrote:

> *afabri* left a comment (CGAL/cgal#9036)
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3306538948>
>
> The ones you see in the testsuite. All we do is to include this header in
> <CGAL/config.h>
> <https://github.com/CGAL/cgal/blob/125c4b254ffc4575f5efec556373806951c826b6/Installation/include/CGAL/config.h#L47>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3306538948>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOE52MOQOEQA7J65LA33TJ5S3AVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMBWGUZTQOJUHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

efifogel committed:
Ops, missed a few min/max occurrences.

efifogel commented:
ops; fixed now; thanks.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Fri, 19 Sept 2025 at 09:17, Andreas Fabri ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Arrangement_on_surface_2/include/CGAL/Arr_linear_traits_2.h
> <https://github.com/CGAL/cgal/pull/9036#discussion_r2361900190>:
>
> > +        const auto* res_pnt = std::get_if<Approx_pnt>(&*result);
> +        CGAL_assertion(res_pnt != nullptr);
> +        *oi++ = *res_pnt;
> +        return oi;
> +      }
> +      if (xcv.is_line()) {
> +        const Line_2 & supp_line = xcv.supp_line();
> +        Approx_lin approx_supp_line(
> +          CGAL::to_double(supp_line.a()),
> +          CGAL::to_double(supp_line.b()),
> +          CGAL::to_double(supp_line.c()));
> +        const auto result = CGAL::intersection(rect, approx_supp_line);
> +        if (! result) return oi;
> +
> +        if (const auto* res_seg = std::get_if<Approx_seg>(&*result)) {
> +          *oi++ = l2r ? res_seg->min() : (res_seg->max)();
>
> You missed those, Efi.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#pullrequestreview-3243304783>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOBOFIEZNRARBINYEKD3TONYLAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTENBTGMYDINZYGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

efifogel commented:
This has been fixed already a few days ago
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Wed, 24 Sept 2025 at 21:17, Sebastien Loriot ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> Arrangement_on_surface_2/doc/Arrangement_on_surface_2/Concepts/AosTraits--ApproximatePoint_2.h
> <https://github.com/CGAL/cgal/pull/9036#discussion_r2376679045>:
>
> > + * \cgalHasModelsBegin
> + * \cgalHasModels{AosApproximatePointTraits_2::Approximate_2}
> + * \cgalHasModels{AosApproximateTraits_2::Approximate_2}
> + * \cgalHasModelsEnd
> + */
> +class ApproximatePoint_2 {
> +public:
> +  /// \name Operations
> +  /// A model of this concept must provide:
> +  /// @{
> +
> +  /*! obtains an approximation of `p`'s \f$x\f$-coordinate (if `i == 0`), or of
> +   * `p`'s \f$y\f$-coordinate (if `i == 1`).
> +   * \pre `i` is either 0 or 1.
> +   */
> +  CGAL::Approximate_number_type operator()(AosTraits::Point_2 p, int i);
>
> @efifogel <https://github.com/efifogel> I think you don't want the CGAL::
> but only Approximate_number_type
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#discussion_r2376679045>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNOCVNQY2RPMOBKFPRLD3ULN2HAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTENRUGEZTOOBRHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sloriot commented:
@efifogel please resolve all conversations

efifogel commented:
I believe that I've addressed all issues:
1. The `CGAL::" prefix was removed (`CGAL::Approximate_number_type`) some
time ago.
2. Regarding the question of whether the `CGAL::Approximate_number_type`
(now `Approximate_number_type`) type definition is needed at all; the
answer is yes.
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/




On Fri, 17 Oct 2025 at 11:32, Sebastien Loriot ***@***.***>
wrote:

> *sloriot* left a comment (CGAL/cgal#9036)
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3414425663>
>
> @efifogel <https://github.com/efifogel> please resolve all conversations
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CGAL/cgal/pull/9036#issuecomment-3414425663>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVBNODRSY7KKVKCGE772AD3YCSRTAVCNFSM6AAAAACEJIMON2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIMJUGQZDKNRWGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

efifogel committed:
Merge branch 'master' into Aos_2-approximate_unbounded-efif

efifogel committed:
cleaned up

efifogel committed:
Fixed link to draw() functions

efifogel committed:
Fixed draw() links

efifogel committed:
Fixed link to draw()

efifogel committed:
Fixed link to draw()

efifogel committed:
Cleaned up
